kythe/docs/schema/indexing-protobuf.txt

```diff
@@ -31,12 +31,235 @@ implementation, as this is often difficult (or uninteresting) to read. This
 document describes how to encode information about generated code to permit
 cross-language links.
 
+To make the discussion easier to understand let's pretend we are working with
+two languages: AliceLang and BobLang. AliceLang has `.alice` file and BobLang
+has `.bob` files. We also have a tool (generator) that can take generate
+`foo.bob` file from `foo.alice` file. We have following components:
+
+* Alice Indexer - Kythe indexer that takes `.alice` files and outputs index
+  data.
+* Bob Indexer - Kythe indexer that takes `.bob` files and outputs index
+  data.
+* Generator - tool that produces `.bob` files from `.alice` files.
+* Post processor - Kythe tool that takes all index data produced by all
+  indexers, processes it and outputs final Kythe graph that contains data
+  for both AliceLang and BobLang.
+
+Now we want to teach Kythe how to create cross-references between generated
+`foo.bob` file and original `foo.alice` file. The main idea is pretty simple:
+Generator has to output extra data containing mapping of elements in `foo.bob`
+to the original elements from `alice.bob`. Then when Bob Indexer is indexing
+`foo.bob` it will use that mapping to output *generates* or *imputes* edges.
+These edges connect nodes from `foo.bob` with nodes in `foo.alice`.
+
+Kythe doesn't require implementors to use one concrete approach for passing
+mapping metadata and outputting *generates* and *imputes* edges. Below we
+describe two different approaches, each has its own pros and cons. But in
+both cases it is assumed that implementors can change Generator and Bob
+Indexer.
+
 TIP: You can find an example implementation at
 link:https://github.com/kythe/kythe/tree/master/kythe/examples/proto[GitHub].
 The current sample web UI does not interpret the parts of the schema we will
 use; this is a work in progress.
 
-== Protocol Buffers
+=== Java To JavaScript with *imputes* edges
+
+This approach is generic and works for any combination of AliceLang and
+BobLang. In this example we generate JavaScript files from Java file so
+AliceLang is Java and BobLang is JavaScript. Given `Color.java`:
+
+[source,java]
+-------------------------------------------------------------------------------
+public enum Color {
+  RED;
+}
+-------------------------------------------------------------------------------
+
+Generator produces `color.js`:
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+-------------------------------------------------------------------------------
+
+==== Changes to Generator
+
+To support cross-references betwen `color.js` and `Color.java` we need to update
+Generator to output the following mapping data for `Color`, `RED` and `GREEN`
+elements.
+
+[source,json]
+-------------------------------------------------------------------------------
+{
+    "type": "kythe0",
```

Comment:
We should have a schema for these values as well, rather than just guessing at the supported bits :-(

Reply:
The purpose of schema would be mostly to remove the need for users to come with their own schema? Because generators and indexers would still need to write code that creates/parses JSON. I don't think Kythe can provide much tools for working with that schema given that Kythe supports many languages and re-implementing these tools in all possible languages as part of Kythe repo seems heavy.

Reply:
Considering that the metadata needs to be understood by the target language indexer and many code generators can target several different languages from a particular source, it seems like there would be a fair amount of benefit from standardizing and documenting a single such format, rather than having each target language use a subtly different one.

Reply:
Ok, I'll address it in a follow up PR.